torchgen/api/translate.py

```diff
@@ -337,10 +337,16 @@ def direct_solve(goal: NamedCType) -> str:
                 )
                 return f"c10::asIntArrayRefSlow({symIntArrayRef_type})"
         elif goal.type == BaseCType(symIntArrayRefT):
-            return direct_solve(NamedCType(goal.name, longSymVec_ctype))
+            try:
+                r = direct_solve(NamedCType(goal.name, BaseCType(intArrayRefT)))
```

Comment:
hmm, these cases cover converting `IntArrayRef` -> `SymIntArrayRef` and `vector<SymInt>` -> `SymIntArrayRef`

Thinking out loud - do we need `vector<int>` -> `SymIntArrayRef` to work too? Maybe not, if tests are passing.

Reply:
I'm actually not sure that we need the `vector<SymInt>` conversion at all; I kept it here because the old code had it.

Reply:
Confirmed needed, it's triggered by your code:

```
  File "/raid/ezyang/pytorch-scratch2/torchgen/gen_functionalization_type.py", line 716, in gen_functionalization_definition
    view_defs.append(emit_view_functionalization_body(g, view_inplace=False))
  File "/raid/ezyang/pytorch-scratch2/torchgen/gen_functionalization_type.py", line 387, in emit_view_functionalization_body
    return {forward_lambda.inner_call(reapply_views=True)}
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/types.py", line 739, in inner_call
    e.expr for e in translate.translate(full_ctx, call_bindings, method=False)
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 405, in translate
    return [Expr(solve(g, direct=False), g) for g in goal_ctypes]
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 405, in <listcomp>
    return [Expr(solve(g, direct=False), g) for g in goal_ctypes]
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 340, in solve
    r = direct_solve(NamedCType(goal.name, BaseCType(intArrayRefT)))
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 235, in direct_solve
    return solve(goal, direct=True)
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 261, in solve
    unsat(goal)
  File "/raid/ezyang/pytorch-scratch2/torchgen/api/translate.py", line 214, in unsat
    raise UnsatError(
torchgen.api.translate.UnsatError: 
Failed to synthesize the expression "at::IntArrayRef size".
When I failed, the following bindings were available in the context:

  const at::Tensor & base; // base
  int64_t mutated_view_idx; // mutated_view_idx
  bool reapply_views; // reapply_views
  ::std::vector<c10::SymInt> size; // size
  bool implicit; // implicit

This probably means there is a missing rule in the rules of torchgen.api.translate.
Check this module for more information.

  in native_functions.yaml line /raid/ezyang/pytorch-scratch2/cmake/../aten/src/ATen/native/native_functions.yaml:2272:
    expand(Tensor(a) self, SymInt[] size, *, bool implicit=False) -> Tensor(a)
```

Reply:
ah yeah, it's for when we need to save `sizes()` in a lambda somewhere (which we need for functionalization to do the view replays).